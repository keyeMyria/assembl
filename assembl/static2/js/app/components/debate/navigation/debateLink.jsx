// @flow
import React from 'react';
import { Link } from 'react-router';
import classNames from 'classnames';
import { I18n } from 'react-redux-i18n';

import Timeline, { type PhaseData } from './timeline';

type DebateLinkProps = {
  identifier: string,
  className: string,
  activeClassName: string,
  children: Array<*>,
  to: string,
  dataText: string,
  screenTooSmall: boolean
};

type DebateLinkState = {
  timelineActive: boolean,
  activePhaseMenu: PhaseData
};

class DebateLink extends React.Component<*, DebateLinkProps, DebateLinkState> {
  state = {
    timelineActive: false,
    activePhaseMenu: { id: '', title: '' }
  };

  timeline = null;

  showTimeline = (event: SyntheticMouseEvent<>) => {
    this.setState({ timelineActive: true });
    if (this.props.screenTooSmall) event.preventDefault();
  };

  hideTimeline = () => {
    this.setState({ timelineActive: false });
  };

  onMenuShown = (phase: PhaseData) => {
    this.setState({ activePhaseMenu: phase });
  };

  onMenuClosed = () => {
    this.setState({ activePhaseMenu: { id: '', title: '' } });
  };

  closeMenu = () => {
    const { activePhaseMenu } = this.state;
    const menuActive = !!activePhaseMenu.id;
    this.setState({ timelineActive: menuActive, activePhaseMenu: { id: '', title: '' } }, () => {
      if (menuActive && this.timeline) this.timeline.getWrappedInstance().closeMenu();
    });
  };

  render() {
    const { identifier, children, to, className, activeClassName, dataText, screenTooSmall } = this.props;
    const { timelineActive, activePhaseMenu } = this.state;
    const currentPhaseTitle = activePhaseMenu.title;
    return (
      <div
        className={classNames('debate-link', { active: timelineActive })}
        onMouseOver={!screenTooSmall && this.showTimeline}
        onMouseLeave={!screenTooSmall && this.hideTimeline}
      >
        {screenTooSmall && (
          <div className="menu-back-container">
            {timelineActive && (
              <div onClick={this.closeMenu} className="menu-back-btn">
                <span className="assembl-icon assembl-icon-left-small" />
                <span className="title">{I18n.t('debate.menuBack')}</span>
              </div>
            )}
          </div>
        )}
        <Link
          to={to}
          className={classNames('debate-link-title', className, { selected: timelineActive })}
          activeClassName={activeClassName}
          data-text={dataText}
        >
          <span className={screenTooSmall && 'title-container'}>
            {children}
            {screenTooSmall &&
              currentPhaseTitle && (
                <span className="phase-title">
                  <span className="assembl-icon assembl-icon-right-dir" />
                  <span>{currentPhaseTitle}</span>
                </span>
              )}
          </span>
          {screenTooSmall &&
            !timelineActive && <span onClick={this.showTimeline} className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
        </Link>
        <div className={classNames('header-container', { integreted: screenTooSmall })}>
          <section className={classNames('timeline-section', { vertical: screenTooSmall })} id="timeline">
            <div className="max-container">
              <Timeline
                ref={(timeline) => {
                  this.timeline = timeline;
                }}
                vertical={screenTooSmall}
                showNavigation
                identifier={identifier}
                onMenuItemClick={this.hideTimeline}
                onMenuShown={screenTooSmall && this.onMenuShown}
                onMenuClosed={screenTooSmall && this.onMenuClosed}
              />
            </div>
          </section>
        </div>
      </div>
    );
  }
}

export default DebateLink;