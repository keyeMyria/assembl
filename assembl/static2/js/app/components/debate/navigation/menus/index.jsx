// @flow
import React from 'react';
import TabMenuList from './tableMenuList';
import IndexMenuList from './indexMenuList';

export const MENU_KINDS = {
  index: 'index',
  table: 'table'
};

export type ItemNode = {
  id: string,
  title: string,
  img: {
    externalUrl: string
  },
  numContributors: number,
  numPosts: number,
  parentId: string
};

type MenuListProps = {
  kind: string
};

function MenuList(props: MenuListProps) {
  switch (props.kind) {
  case MENU_KINDS.index:
    return <IndexMenuList {...props} />;
  default:
    return <TabMenuList {...props} />;
  }
}

export { default as SurveyMenu } from './surveyMenu';
export { default as IdeasMenu } from './ideasMenu';
export { default as Menu } from './menu';
export default MenuList;