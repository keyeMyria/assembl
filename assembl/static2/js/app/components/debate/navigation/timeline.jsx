// @flow
import React from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { isCurrentPhase, getBarPercent, isStepCompleted } from '../../../utils/timeline';
import TimelineSegment, { type DebateType } from './timelineSegment';

export type PhaseData = {
  id: string,
  title: string
};

type TimelineProps = {
  debate: DebateType,
  showNavigation: boolean,
  identifier: string,
  vertical: boolean,
  onMenuItemClick: Function,
  onMenuShown: Function,
  onMenuClosed: Function
};

type TimelineState = {
  selectedPhase: ?string
};

export class DumbTimeline extends React.Component<*, TimelineProps, TimelineState> {
  state = {
    selectedPhase: null
  };

  segments = {};

  onShowMenu = (phase: PhaseData) => {
    const { onMenuShown } = this.props;
    this.setState({ selectedPhase: phase.id }, () => {
      if (onMenuShown) onMenuShown(phase);
    });
  };

  onHideMenu = () => {
    const { onMenuClosed } = this.props;
    this.setState({ selectedPhase: null }, () => {
      if (onMenuClosed) onMenuClosed();
    });
  };

  closeMenu = () => {
    const { selectedPhase } = this.state;
    const segment = selectedPhase && this.segments[selectedPhase];
    if (segment) {
      segment
        // connec wrapper
        .getWrappedInstance()
        // apollo wrapper
        .getWrappedInstance()
        .close();
    }
    this.setState({ selectedPhase: null });
  };

  render() {
    const { debateData } = this.props.debate;
    const { showNavigation, identifier, onMenuItemClick, vertical } = this.props;
    const { selectedPhase } = this.state;
    return (
      <div
        className={classNames(
          'timeline-container',
          { vertical: vertical },
          { 'timeline-menu-container': vertical && selectedPhase }
        )}
      >
        {debateData.timeline &&
          debateData.timeline.map((phase, index) => {
            if (!vertical || !selectedPhase || selectedPhase === phase.identifier) {
              return (
                <TimelineSegment
                  ref={(segment) => {
                    this.segments[phase.identifier] = segment;
                  }}
                  vertical={vertical}
                  title={phase.title}
                  index={index}
                  key={index}
                  barPercent={getBarPercent(debateData.timeline[index])}
                  isCurrentPhase={isCurrentPhase(debateData.timeline[index])}
                  showNavigation={showNavigation}
                  identifier={identifier}
                  phaseIdentifier={phase.identifier}
                  startDate={phase.start}
                  endDate={phase.end}
                  isStepCompleted={isStepCompleted(phase)}
                  onMenuItemClick={onMenuItemClick}
                  onShowMenu={vertical && this.onShowMenu}
                  onHideMenu={vertical && this.onHideMenu}
                />
              );
            }
            return null;
          })}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  debate: state.debate
});

export default connect(mapStateToProps, null, null, { withRef: true })(DumbTimeline);