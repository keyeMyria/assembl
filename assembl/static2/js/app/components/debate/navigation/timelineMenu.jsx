// @flow
import React from 'react';
import { withApollo } from 'react-apollo';
import { Translate, Localize } from 'react-redux-i18n';
import { connect } from 'react-redux';

import Menu from './menus/menu';
import { getPhaseStatus, type Phase } from '../../../utils/timeline';
import { PHASE_STATUS } from '../../../constants';
import { MENU_KINDS } from './menus';

type TimelineMenuProps = {
  phase: Phase,
  locale: string,
  vertical: boolean,
  onMenuItemClick: Function
};

export class DumbTimelineMenu extends React.Component<*, TimelineMenuProps, void> {
  componentWillMount() {
    const { phase: { title, start, end }, locale } = this.props;
    this.phaseStatus = getPhaseStatus(start, end);
    let phaseName = '';
    title.entries.forEach((entry) => {
      if (locale === entry['@language']) {
        phaseName = entry.value.toLowerCase();
      }
    });
    this.phaseName = phaseName;
  }

  phaseStatus = null;

  phaseName = null;

  renderNotStarted = (className?: string) => {
    const { phase: { start } } = this.props;
    return (
      <div className={className}>
        <Translate value="debate.notStarted" phaseName={this.phaseName} />
        <Localize value={start} dateFormat="date.format" />
      </div>
    );
  };

  render() {
    const { phase: { identifier }, onMenuItemClick, vertical } = this.props;
    const isNotStarted = this.phaseStatus === PHASE_STATUS.notStarted;
    return (
      <div className="menu-container">
        {isNotStarted ? (
          this.renderNotStarted('not-started')
        ) : (
          <Menu identifier={identifier} onMenuItemClick={onMenuItemClick} kind={vertical ? MENU_KINDS.index : MENU_KINDS.table} />
        )}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale
});

// $FlowFixMe
export default connect(mapStateToProps)(withApollo(DumbTimelineMenu));