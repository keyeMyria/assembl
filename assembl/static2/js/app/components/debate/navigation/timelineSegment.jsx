// @flow
import React from 'react';
import { withApollo, type ApolloClient } from 'react-apollo';
import { browserHistory } from 'react-router';
import { Translate, Localize, I18n } from 'react-redux-i18n';
import { connect } from 'react-redux';
import classNames from 'classnames';

import Menu, { prefetchMenuQuery } from './menus/menu';
import { getPhaseStatus, isSeveralIdentifiers, type Timeline } from '../../../utils/timeline';
import { displayModal } from '../../../utils/utilityManager';
import { get } from '../../../utils/routeMap';
import { PHASE_STATUS } from '../../../constants';
import { MENU_KINDS } from './menus';

export type DebateType = {
  debateData: {
    timeline: Timeline,
    slug: string
  }
};

type TimelineSegmentProps = {
  index: number,
  client: ApolloClient,
  title: {
    entries: Array<*>
  },
  startDate: string,
  endDate: string,
  phaseIdentifier: string,
  debate: DebateType,
  barPercent: number,
  isCurrentPhase: boolean,
  isStepCompleted: boolean,
  locale: string,
  vertical: boolean,
  onMenuItemClick: Function,
  onShowMenu: Function,
  onHideMenu: Function
};

type TimelineSegmentState = {
  active: boolean
};

export class DumbTimelineSegment extends React.Component<*, TimelineSegmentProps, TimelineSegmentState> {
  state = {
    active: false
  };

  componentWillMount() {
    const { phaseIdentifier, title, startDate, endDate, locale, client } = this.props;
    this.phaseStatus = getPhaseStatus(startDate, endDate);
    let phaseName = '';
    let phaseTitle = '';
    title.entries.forEach((entry) => {
      if (locale === entry['@language']) {
        phaseTitle = entry.value;
        phaseName = entry.value.toLowerCase();
      }
    });
    this.phaseName = phaseName;
    this.phaseTitle = phaseTitle;
    prefetchMenuQuery(client, {
      lang: locale,
      identifier: phaseIdentifier
    });
  }

  componentWillReceiveProps() {
    if (this.state.active) {
      this.setState({ active: false });
    }
  }

  phaseStatus = null;

  phaseName = null;

  phaseTitle = null;

  showMenu = (event: SyntheticMouseEvent<>) => {
    this.setState({ active: true }, () => {
      const { onShowMenu, phaseIdentifier } = this.props;
      if (onShowMenu) onShowMenu({ id: phaseIdentifier, title: this.phaseTitle });
    });
    if (this.props.vertical && event) event.preventDefault();
  };

  hideMenu = () => {
    this.setState({ active: false }, () => {
      const { onHideMenu } = this.props;
      if (onHideMenu) onHideMenu();
    });
  };

  close = () => {
    this.setState({ active: false });
  };

  renderNotStarted = (className?: string) => {
    const { startDate } = this.props;
    return (
      <div className={className}>
        <Translate value="debate.notStarted" phaseName={this.phaseName} />
        <Localize value={startDate} dateFormat="date.format" />
      </div>
    );
  };

  displayPhase = () => {
    const { phaseIdentifier } = this.props;
    const { debateData } = this.props.debate;
    const phase = debateData.timeline.filter(p => p.identifier === phaseIdentifier);
    const isRedirectionToV1 = phase[0].interface_v1;
    const slug = { slug: debateData.slug };
    const params = { slug: debateData.slug, phase: phaseIdentifier };
    const isSeveralPhases = isSeveralIdentifiers(debateData.timeline);
    if (isSeveralPhases) {
      if (this.phaseStatus === PHASE_STATUS.notStarted) {
        const body = this.renderNotStarted();
        displayModal(null, body, true, null, null, true);
      }
      if (this.phaseStatus === PHASE_STATUS.inProgress || this.phaseStatus === PHASE_STATUS.completed) {
        if (!isRedirectionToV1) {
          browserHistory.push(get('debate', params));
          this.hideMenu();
        } else {
          const body = <Translate value="redirectToV1" phaseName={this.phaseName} />;
          const button = { link: get('oldDebate', slug), label: I18n.t('home.accessButton'), internalLink: false };
          displayModal(null, body, true, null, button, true);
          setTimeout(() => {
            window.location = get('oldDebate', slug);
          }, 6000);
        }
      }
    } else if (!isRedirectionToV1) {
      browserHistory.push(get('debate', params));
      this.hideMenu();
    } else {
      window.location = get('oldDebate', slug);
    }
  };

  renderMenu = () => {
    const { phaseIdentifier, onMenuItemClick, vertical } = this.props;
    const { active } = this.state;
    const isNotStarted = this.phaseStatus === PHASE_STATUS.notStarted;
    if (!active) return null;
    return (
      <div className="menu-container">
        {isNotStarted ? (
          this.renderNotStarted('not-started')
        ) : (
          <Menu
            identifier={phaseIdentifier}
            onMenuItemClick={onMenuItemClick}
            kind={vertical ? MENU_KINDS.index : MENU_KINDS.table}
          />
        )}
      </div>
    );
  };

  render() {
    const { index, barPercent, isCurrentPhase, isStepCompleted, vertical } = this.props;
    const { active } = this.state;
    const timelineClass = classNames('timeline-title', {
      'txt-active-bold': vertical && isCurrentPhase,
      'txt-active-light': !vertical || (vertical && isStepCompleted),
      'txt-not-active': vertical && !isCurrentPhase && !isStepCompleted
    });
    const complitedOrCurrent = isStepCompleted || isCurrentPhase;
    const dimensionName = vertical ? 'height' : 'width';
    if (vertical && active) return this.renderMenu();
    return (
      <div
        className={classNames('minimized-timeline', {
          active: active
        })}
        onMouseOver={!vertical && this.showMenu}
        onMouseLeave={!vertical && this.hideMenu}
        onClick={vertical && this.showMenu}
      >
        <div onClick={this.displayPhase} className={timelineClass}>
          <div className="timeline-link">{this.phaseTitle}</div>
          {vertical && (
            <div className="timeline-link-description">
              {isStepCompleted && I18n.t('debate.phaseCompleted')}
              {isCurrentPhase && I18n.t('debate.phaseStarted')}
              {!complitedOrCurrent && I18n.t('debate.phaseUpcoming')}
            </div>
          )}
        </div>
        <div className="timeline-graph">
          {vertical && (
            <div
              className={classNames('timeline-number', {
                active: complitedOrCurrent,
                'not-active': !complitedOrCurrent,
                'timeline-text': !isStepCompleted
              })}
            >
              {isStepCompleted ? (
                <span className="assembl-icon-checked white" />
              ) : (
                <span>
                  {isCurrentPhase && <span className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
                  {index + 1}
                </span>
              )}
            </div>
          )}
          <div className="timeline-bars">
            {barPercent > 0 && (
              <div
                className="timeline-bar-filler"
                style={barPercent < 20 ? { [dimensionName]: '20%' } : { [dimensionName]: `${barPercent}%` }}
              >
                &nbsp;
              </div>
            )}
            <div className="timeline-bar-background">&nbsp;</div>
          </div>
        </div>
        {!vertical && active && <span className="timeline-arrow" />}
        {this.renderMenu()}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale,
  debate: state.debate
});

// $FlowFixMe
export default connect(mapStateToProps, null, null, { withRef: true })(withApollo(DumbTimelineSegment, { withRef: true }));