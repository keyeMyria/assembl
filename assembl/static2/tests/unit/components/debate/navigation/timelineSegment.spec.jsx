import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTimelineSegment } from '../../../../../js/app/components/debate/navigation/timelineSegment';

describe('DumbTimeline component', () => {
  it('should match the DumbTimeline', () => {
    const timeline = [
      {
        identifier: 'foo',
        start: 'date1',
        end: 'date2',
        title: { entries: [{ en: 'Foo' }] }
      },
      {
        identifier: 'bar',
        start: 'date1Bar',
        end: 'date2Bar',
        title: { entries: [{ en: 'Bar' }] }
      }
    ];
    const phase = {
      title: {
        entries: [{ en: 'Foo' }]
      },
      start: 'date1',
      end: 'date2',
      identifier: 'foo'
    };
    const props = {
      phase: phase,
      client: { query: () => {} },
      barPercent: 20,
      locale: 'en',
      debate: {
        debateData: {
          slug: 'slug',
          timeline: timeline
        }
      }
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTimelineSegment {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});