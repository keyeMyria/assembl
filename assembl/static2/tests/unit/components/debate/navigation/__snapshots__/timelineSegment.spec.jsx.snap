// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbTimeline component should match the DumbTimeline 1`] = `
<div
  className="minimized-timeline"
  onClick={undefined}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    className="timeline-title txt-active-light"
    onClick={[Function]}
  >
    <div
      className="timeline-link"
    />
  </div>
  <div
    className="timeline-graph"
  >
    <div
      className="timeline-bars"
    >
      <div
        className="timeline-bar-filler"
        style={
          Object {
            "width": "20%",
          }
        }
      >
         
      </div>
      <div
        className="timeline-bar-background"
      >
         
      </div>
    </div>
  </div>
</div>
`;
